39046 va-date Web Component

pdavies88 opened the pull request:
## Chromatic
<!-- This `39046-date-web-component` is a placeholder for a CI job - it will be updated automatically -->
https://39046-date-web-component--60f9b557105290003b387cd5.chromatic.com

## Description
Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/39046

## Notes
This should not be merged in prior to the completion of `va-number-input`

## Testing done
<img width="483" alt="Screen Shot 2022-04-13 at 9 46 26 AM" src="https://user-images.githubusercontent.com/11822533/163195140-ba568002-859f-4735-a192-8a2919fb5267.png">

## Screenshots
<img width="1440" alt="Screen Shot 2022-04-13 at 9 46 37 AM" src="https://user-images.githubusercontent.com/11822533/163195156-6d79af0d-5338-4d65-a864-598da2b9e1b9.png">


## Acceptance criteria
- [X] `va-date` is written and added to Storybook
- [X] E2E tests are written
- [ ] Add Chromatic link to #39047, assign ticket to Courtney, and move ticket to Current Sprint column
- [ ] Once accessibility review is complete, alert Carol that component is ready for staging review
- [ ] Address any staging review comments

## Definition of done
- [ ] Documentation has been updated, if applicable
- [X] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

k80bowman commented:
This looks good to me. The only comment I have is that, when using voiceover on the Mac, it reads the first option for the month and day as "new line". It does this for the React component as well, so I'm not going to hold up the PR for this, but it might be nice to figure out a better option. I tried to do some research on how other design system date components handle this, but it seems that most other date components either only use text input boxes or are date pickers with a calendar attached.

bkjohnson commented on packages/storybook/stories/va-select.stories.jsx:
```diff
@@ -77,7 +75,6 @@ const Template = ({
         value={value}
         required={required}
         error={error}
-        aria-live-region-text={ariaLiveRegionText}
```
Is this related to date things?

pdavies88 replied:
It was not in use and seems like we forgot to remove it.

pdavies88 replied:
I will undo the change and write up a ticket for it instead

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,242 @@
+  /**
+   * Label for the field.
+   */
+  @Prop() label!: string;
```
I like the `!` option - hopefully it will make storybook more useful.

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,242 @@
+   */
+  @Prop() maxYear: number;
+
+  /**
+   * Set the default month on the month input.
+   */
+  @Prop() month: string;
+
+  /**
+   * Set the default day on the day input.
+   */
+  @Prop() day: string;
+
+  /**
+   * Set the default year on the year input.
+   */
+  @Prop() year: string;
```
Can we also get a `value` prop since that's an established pattern for inputs? That would also let us use the familiar `event.target.value` pattern.

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,242 @@
+    composed: true,
+    bubbles: true,
+  })
+  monthBlurEvent: EventEmitter;
+
+  /**
+   * Fires when the day input loses focus after its value was changed
+   */
+  @Event({
+    composed: true,
+    bubbles: true,
+  })
+  dayChangeEvent: EventEmitter;
+
+  /**
+   * Fires when the day input loses focus
+   */
+  @Event({
+    composed: true,
+    bubbles: true,
+  })
+  dayBlurEvent: EventEmitter;
+
+  /**
+   * Fires when the year input loses focus after its value was changed
+   */
+  @Event({
+    composed: true,
+    bubbles: true,
+  })
+  yearChangeEvent: EventEmitter;
+
+  /**
+   * Fires when the year input loses focus
+   */
+  @Event({
+    composed: true,
+    bubbles: true,
+  })
+  yearBlurEvent: EventEmitter;
```
Are these custom events necessary? Also can we get a `dateChange` event or something? The current `Date` react component has the single event handler:

https://github.com/department-of-veterans-affairs/component-library/blob/8f70616a228f1a1a6d2f0c6980c0c1ac402e1c35/packages/react-components/src/components/Date/Date.jsx#L54

bkjohnson reviewed (commented):
Some questions about the events.

pdavies88 commented on packages/react-components/src/helpers/options-for-select.js:
```diff
@@ -1,43 +1,3 @@
-export const maritalStatuses = [
```
All of these are unused in Component Library and were copy and pasted from vets-website. Removing since it is unused code.

pdavies88 commented:
@bkjohnson @k80bowman @andresriveratoro this component has been reworked to only use a singular blur and change event for the component. It also takes in a singular prop for the date named value and value must be formatted as YYYY-MM-DD to behave correctly. 

Please take a look again and let me know what other steps would need to be taken to complete the development work for this component prior to a staging review.

k80bowman commented:
The updates look great, thank you @pdavies88. I do have one unrelated question, is the year input supposed to have these up and down buttons? I don't see them in the React component:
![image](https://user-images.githubusercontent.com/12739849/164059930-8d8beca9-00b1-4a93-a7ae-b5b49f2d4628.png)

pdavies88 commented:
@k80bowman they are there by default but CSS can be used to overwrite their appearance: https://www.w3schools.com/howto/howto_css_hide_arrow_number.asp
In the React Component they are being hidden via this code:
```
.usa-date-of-birth [type=number] {
    -moz-appearance: textfield;
}
```
I will update the CSS of the `va-number-input` to include that code so they no longer appear.

k80bowman reviewed (approved):
This looks good to me, thanks for all your work on this. And especially for adding the documentation around validation, that's great.

bkjohnson reviewed (commented):
Mostly some comments about the interface for other engineers.

pdavies88 commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,207 @@
+
+    let { month, day, year } = this;
+
+    const [defaultYear, defaultMonth, defaultDay] = (value || '').split('-');
```
I was running into console errors when trying to set a default value so I decided to leave as is

pdavies88 commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,207 @@
+      ? days[month]
+      : parseInt(defaultMonth, 10) > 0
+      ? days[parseInt(defaultMonth, 10)]
+      : [];
```
Comment has been added and code was condensed

pdavies88 commented:
@k80bowman @andresriveratoro @bkjohnson I have updated and reworked this PR again based on Brooks most recent comments. Please re-review the PR and let me know if additional changes are needed

pdavies88 commented on packages/storybook/stories/va-date.stories.jsx:
```diff
@@ -0,0 +1,211 @@
+        min-year={minYear}
+        max-year={maxYear}
+        onDateBlurEvent={e => console.log(e, 'DATE BLUR EVENT')}
+        onDateChangeEvent={e => setMonthDayYear(e.detail.path[0])}
```
Code has been reworked to allow for just e.target.value to be used

pdavies88 commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,207 @@
+  /**
+   * Set the default date value must be in YYYY-MM-DD format.
+   */
+  @Prop() value: string;
```
Fixed

pdavies88 commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,207 @@
+
+    let newDate = `${year || defaultYear}-${
+      (month && this.minTwoDigits(month)) || defaultMonth
+    }-${(day && this.minTwoDigits(day)) || defaultDay}`;
```
Is have updated to use this thanks

pdavies88 commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,207 @@
+    if (target.classList.contains('input-year')) {
+      this.year = target.value;
+    }
+    this.dateChangeEvent.emit(event);
```
I have left a comment in the code and left the event firing as it currently does. Mainly it starts us on the path of controlling what is a valid date which might differ from team to team and I would like the app teams to control the validation aspect instead of our team.

pdavies88 commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,207 @@
+    composed: true,
+    bubbles: true,
+  })
+  dateBlurEvent: EventEmitter;
```
I have updated to remove the word event

pdavies88 commented on packages/storybook/stories/va-date.stories.jsx:
```diff
@@ -0,0 +1,211 @@
+    if (target.tagName === 'INPUT') {
+      setYearVal(target.value);
+    }
+  };
```
Code has been reworked to allow for just e.target.value to be used

k80bowman commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -5,6 +5,7 @@ import {
   days,
 } from '../../../../react-components/src/helpers/options-for-select.js';
 
+import { isFullDate } from '../../../../react-components/src/helpers/validations.js';
```
We probably shouldn't be importing things like this from the react components. That could cause problems when we go to delete the React component, if whoever is doing that work doesn't know these imports are here. Could we copy these imports to a `date-utils.ts` file in the web components `utils` directory? Sorry I overlooked the months and days imports in my initial review.

pdavies88 replied:
@k80bowman updates have been made and the `date-utils.ts` file has been created

k80bowman commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,208 @@
+      parseInt(month, 10) > 0 ? days[parseInt(month, 10)] : [];
+
+    // Check validity of date if invalid provide message and error state styling
+    const dateInvalid =
+      required && (!isFullDate(value) || day > daysForSelectedMonth.length)
+        ? 'Please provide a valid date'
+        : null;
```
Thank you for adding this in. Could you please add a test for it?

pdavies88 replied:
@k80bowman yes testing suite has been updated

k80bowman commented on packages/web-components/src/utils/date-utils.ts:
```diff
@@ -0,0 +1,130 @@
+  12: thirtyOneDays,
+};
+
+export function isFullDate(date) {
+  return /\d{4}-\d{1,2}-\d{1,2}/.test(date);
+}
```
Thank you for copying all of this over. I know there isn't a test for this in the React components directory, but I'm wondering if maybe there should be here.

pdavies88 replied:
@k80bowman yes testing suite has been updated

bkjohnson commented on packages/web-components/src/components/va-date/test/va-date.e2e.ts:
```diff
@@ -0,0 +1,147 @@
+import { newE2EPage } from '@stencil/core/testing';
+import { axeCheck } from '../../../testing/test-helpers';
+
+describe('va-date', () => {
```
Can we also get a test showing that the event target's `value` has the updated date for the `dateChange` event?

pdavies88 replied:
I added to the test where I check for updates: https://github.com/department-of-veterans-affairs/component-library/pull/347/commits/601be1531e3ee837c5397954467269da578a0613#diff-999e0d6b4f145596771d22dcd3d694a5363c5f85675d8a905c4987e89c1cf631

bkjohnson replied:
I still don't think we have a test for `dateChange` - I think it would be good to have one to test that it's firing when any 1 of the three components has a change.

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,208 @@
+      value,
+    } = this;
+
+    const [year, month, day] = (value || '').split('-');
```
After the split we could probably do a `.map()` which parses each value as an int.

pdavies88 replied:
Change has been completed

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,208 @@
+
+    // Convert string to number to remove leading 0 when matching to days object
+    const daysForSelectedMonth =
+      parseInt(month, 10) > 0 ? days[parseInt(month, 10)] : [];
```
And that would simplify this.

pdavies88 replied:
Change has been completed

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,208 @@
+              label="Month"
+              name={`${name}Month`}
+              // Remove starting 0 on days less than 10
+              value={month ? month.replace(/^0+/, '') : ''}
```
And maybe this as well

pdavies88 replied:
Change has been completed

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,208 @@
+                daysForSelectedMonth.length < day
+                  ? ''
+                  : day
+                  ? day.replace(/^0+/, '')
```
Here too.

pdavies88 replied:
Change has been completed

bkjohnson reviewed (commented):
I think it would be good to have some tests for how the `dateChange` event works.

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -154,8 +155,8 @@ export class VaDate {
-              // Remove starting 0 on days less than 10
-              value={month ? month.replace(/^0+/, '') : ''}
+              // Value must be a string
+              value={month ? month.toString() : ''}
```
Would `month?.toString()` work instead of having the ternary?

pdavies88 replied:
Change has been completed

pdavies88 commented:
@bkjohnson e2e tests to for the dateBlur and dateChange events have been added to the PR

bkjohnson reviewed (commented):
Adding a `required` part to the select and number input so that we can set `display:none` when used inside of `<va-date>` means that others can do the same thing which I'm not too keen on. Is there a reason why we are doing that?

bkjohnson commented on packages/web-components/src/components/va-number-input/va-number-input.tsx:
```diff
@@ -112,6 +116,8 @@ export class VaNumberInput {
           onInput={this.handleInput}
           onBlur={this.handleBlur}
+          part="input"
+          required={this.required || null}
```
Is this meant to be left in?

bkjohnson commented on packages/web-components/src/components/va-select/va-select.tsx:
```diff
@@ -136,6 +136,8 @@ export class VaSelect {
           onKeyDown={() => this.handleKeyDown()}
           onChange={e => this.handleChange(e)}
+          part="select"
+          required={required || null}
```
Here as well.

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,206 @@
+    // Fieldset has an implicit aria role of group
+    return (
+      <Host value={value} invalid={dateInvalid}>
+        <fieldset aria-label="Select Month and two digit day XX and four digit year format XXX">
```
@k80bowman we'll need to make sure we get a translation for this as well since this is new.

k80bowman replied:
Good call, I'll add it to the spreadsheet and alert the check-in team. Thanks!

bkjohnson replied:
We're also [missing an X](https://github.com/department-of-veterans-affairs/va.gov-team/issues/39047#issuecomment-1104468838).

```suggestion
        <fieldset aria-label="Select Month and two digit day XX and four digit year format XXXX">
```

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,206 @@
+      <Host value={value} invalid={dateInvalid}>
+        <fieldset aria-label="Select Month and two digit day XX and four digit year format XXX">
+          <legend>
+            {label || 'Date of birth'}{' '}
```
I know that the `Date of birth` thing is what's happening in the React component, but I don't know that there's an actual reason for it. It's just [been that way for over 3 years](https://github.com/department-of-veterans-affairs/component-library/blame/40e0142e0d28eb9a3d3556ffaac652595629babc/src/components/ErrorableDate/ErrorableDate.jsx#L89). Since the uses of `<Date>` and `<SimpleDate>` in `vets-website` are all passing label props, I think it might be good if we remove this rather than make an assumption that this component is for collecting date of birth.

bkjohnson commented on packages/web-components/src/components/va-date/va-date.tsx:
```diff
@@ -0,0 +1,206 @@
+    // Error attribute should be leveraged for custom error messaging
+    // Fieldset has an implicit aria role of group
+    return (
+      <Host value={value} invalid={dateInvalid}>
```
Can you talk a bit more about the possible conflicts with `error` and `invalid`? I'm not sure why we have both.
